pkg/stanza/fileconsumer/file_threadpool.go

```diff
@@ -0,0 +1,198 @@
+	// This is only applicable for `filelog.useThreadPool` featuregate
+	ctx, cancel := context.WithCancel(context.Background())
+	m.syncLastPollFilesConcurrent(ctx)
+	cancel()
+}
+
+// poll checks all the watched paths for new entries
+func (m *Manager) pollConcurrent(ctx context.Context) {
+	// Increment the generation on all known readers
+	// This is done here because the next generation is about to start
+	m.knownFilesLock.Lock()
+	for i := 0; i < len(m.knownFiles); i++ {
+		m.knownFiles[i].generation++
+	}
+	m.knownFilesLock.Unlock()
+
+	// Get the list of paths on disk
+	matches, err := m.fileMatcher.MatchFiles()
+	if err != nil {
+		m.Errorf("error finding files: %s", err)
+	}
+	m.consumeConcurrent(ctx, matches)
+	m.clearCurrentFingerprints()
+
+	// Any new files that appear should be consumed entirely
+	m.readerFactory.fromBeginning = true
+	m.syncLastPollFilesConcurrent(ctx)
+}
+
+func (m *Manager) worker(ctx context.Context) {
+	defer m.workerWg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case chanData, ok := <-m.readerChan:
+			if !ok {
+				return
+			}
+			r, fp := chanData.reader, chanData.fp
+			if !m.readToEnd(ctx, r) {
+				// Save off any files that were not fully read or if deleteAfterRead is disabled
+				m.knownFilesLock.Lock()
+				m.knownFiles = append(m.knownFiles, r)
+				m.knownFilesLock.Unlock()
+			}
+			m.removePath(fp)
+		}
+	}
+}
+
+func (m *Manager) makeReaderConcurrent(filePath string) (*reader, *fingerprint.Fingerprint) {
+	fp, file := m.makeFingerprint(filePath)
+	if fp == nil {
+		return nil, nil
+	}
+
+	// check if the current file is already being consumed
+	if m.isCurrentlyConsuming(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+
+	// Exclude any empty fingerprints or duplicate fingerprints to avoid doubling up on copy-truncate files
+	if m.checkDuplicates(fp) {
+		if err := file.Close(); err != nil {
+			m.Errorf("problem closing file", "file", file.Name())
+		}
+		return nil, nil
+	}
+	m.currentFps = append(m.currentFps, fp)
+
+	reader, err := m.newReaderConcurrent(file, fp)
+	if err != nil {
+		m.Errorw("Failed to create reader", zap.Error(err))
+		return nil, nil
+	}
+	return reader, fp
+}
+
+func (m *Manager) consumeConcurrent(ctx context.Context, paths []string) {
+	m.clearOldReadersConcurrent(ctx)
+	for _, path := range paths {
+		reader, fp := m.makeReaderConcurrent(path)
+		if reader != nil {
+			// add path and fingerprint as it's not consuming
+			m.trieLock.Lock()
+			m.trie.Put(fp.FirstBytes)
+			m.trieLock.Unlock()
+			m.readerChan <- readerWrapper{reader: reader, fp: fp}
+		}
+	}
+}
+
+func (m *Manager) isCurrentlyConsuming(fp *fingerprint.Fingerprint) bool {
+	m.trieLock.RLock()
+	defer m.trieLock.RUnlock()
+	return m.trie.HasKey(fp.FirstBytes)
+}
+
+func (m *Manager) removePath(fp *fingerprint.Fingerprint) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	m.trie.Delete(fp.FirstBytes)
+}
```

Comment:
Should we just move the locking into the trie?

Reply:
Well, we can. But I didn't go with it, as I was planning to add Trie in the current model (synchronous) as well in the future.

Reply:
That's a whole different thing, we can discuss it in a different issue.